File: 25-Deck-of-Cards/80-Deck-of-Cards.js
const shuffle = require('lodash/shuffle')

class Card {
    constructor(value, suit) {
        this.value = value
        this.suit = suit
    }

    // changes "console.log" representation
    inspect() {
        return `${this.value} of ${this.suit}`
    }
}

class Deck {
    constructor() {
        const suits = ['Hearts', 'Diamonds', 'Clubs', 'Spades']
        const values = ['A', '2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K']

        this.cards = []
        suits.forEach(suit => {
            values.forEach(value => {
                this.cards.push(new Card(value, suit))
            })
        })
    }

    inspect() {
        return `Deck of ${this.count()} cards`
    }

    count() {
        return this.cards.length
    }

    shuffle() {
        if (this.count() < 52) throw Error("Only full decks can be shuffled")

        this.cards = shuffle(this.cards)
        return this.cards
    }

    _deal(num) {
        if (this.count() === 0) throw Error("All cards have been dealt")
        // or...  (not sure which is better)
        // if (this.count() === 0) throw { name: "ValueError", message: "All cards have been dealt" }

        const actualNum = Math.min(this.count(), num)

        // "splice" changes the array and returns deleted items (perfect!)
        // removes actualNum cards from end of "this.cards" and returns array of removed cards
        return this.cards.splice(-actualNum, actualNum)
    }

    dealCard() {
        return this._deal(1)[0]
    }

    dealHand(num) {
        return this._deal(num)
    }
}


// === tests ===
let d = new Deck()
d.shuffle()

let card = d.dealCard()
console.log(card)

let hand = d.dealHand(5)
console.log(hand)

d.dealHand(46) // deal the rest of the deck
d.dealCard() // test Error

File: 25-Deck-of-Cards/80-Deck-of-Cards.py
from random import shuffle

class Card:
    def __init__(self, value, suit):
        self.value = value
        self.suit = suit

    def __repr__(self):
        return f"{self.value} of {self.suit}"


class Deck:
    def __init__(self):
        suits = ('Hearts', 'Diamonds', 'Clubs', 'Spades')
        values = ('A', '2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K')

        # initialize list of 52 Cards
        
        # nested for loop way
        # self.cards = []
        # for suit in suits:
        #     for value in values:
        #         self.cards.append(Card(suit, value))

        # list comprehension way
        self.cards = [Card(value, suit) for value in values for suit in suits]

    def __repr__(self):
        return f"Deck of {self.count()} cards"

    def count(self):
        return len(self.cards)

    def shuffle(self):
        if self.count() < 52:
            raise ValueError("Only full decks can be shuffled")

        shuffle(self.cards)
        return self
    
    def _deal(self, num):
        if self.count() == 0:
            raise ValueError("All cards have been dealt")

        # get whichever is smaller    
        actual = min([self.count(), num])

        # remove a slice of cards from the end
        cards = self.cards[-actual:]

        # update self.cards
        self.cards = self.cards[:-actual]
        
        return cards

    def deal_card(self):
        return self._deal(1)[0]  # get single value from the list

    def deal_hand(self, num):
        return self._deal(num)
